to_be_ccppized/coords_1d.F90

```diff
@@ -1,152 +1,154 @@
+
+  coords%ifc = ifc
+  coords%mid = mid
+  coords%del = del
+  coords%dst = dst
+  coords%rdel = rdel
+  coords%rdst = rdst
+
+end function new_Coords1D_from_fields
+
+! Constructor if you only have interface coordinates; derives all the other
+! fields.
+function new_Coords1D_from_int(ifc) result(coords)
+  real(r8), USE_CONTIGUOUS intent(in) :: ifc(:,:)
+  type(Coords1D) :: coords
+
+  coords = allocate_coords(size(ifc, 1), size(ifc, 2) - 1)
+
+  coords%ifc = ifc
+  coords%mid = 0.5_r8 * (ifc(:,:coords%d)+ifc(:,2:))
+  coords%del = coords%ifc(:,2:) - coords%ifc(:,:coords%d)
+  coords%dst = coords%mid(:,2:) - coords%mid(:,:coords%d-1)
+  coords%rdel = 1._r8/coords%del
+  coords%rdst = 1._r8/coords%dst
+
+end function new_Coords1D_from_int
+
+! Create a new Coords1D object that is a subsection of some other object,
+! e.g. if you want only the first m coordinates, use d_bnds=[1, m].
+!
+! Originally this used pointers, but it was found to actually be cheaper
+! in practice just to make a copy, especially since pointers can impede
+! optimization.
+function section(self, n_bnds, d_bnds)
+  class(Coords1D), intent(in) :: self
+  integer, intent(in) :: n_bnds(2), d_bnds(2)
+  type(Coords1D) :: section
+
+  section = allocate_coords(n_bnds(2)-n_bnds(1)+1, d_bnds(2)-d_bnds(1)+1)
+
+  ! The following code is intrepreted by nvhpc 24.9 as an allocation on assign
+  ! section%ifc  = self%ifc(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2)+1)
+  ! section%mid  = self%mid(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2))
+  ! section%del  = self%del(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2))
+  ! section%dst  = self%dst(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2)-1)
+  ! section%rdel = self%rdel(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2))
+  ! section%rdst = self%rdst(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2)-1)
+  section%ifc(:,:)  = self%ifc(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2)+1)
+  section%mid(:,:)  = self%mid(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2))
+  section%del(:,:)  = self%del(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2))
+  section%dst(:,:)  = self%dst(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2)-1)
+  section%rdel(:,:) = self%rdel(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2))
+  section%rdst(:,:) = self%rdst(n_bnds(1):n_bnds(2),d_bnds(1):d_bnds(2)-1)
+
+end function section
+
+! Quick utility to get allocate each array with the correct size.
+function allocate_coords(n, d) result(coords)
+  integer, intent(in) :: n, d
+  type(Coords1D) :: coords
+
+  coords%n = n
+  coords%d = d
+
+  allocate(coords%ifc(coords%n,coords%d+1))
+  allocate(coords%mid(coords%n,coords%d))
+  allocate(coords%del(coords%n,coords%d))
+  allocate(coords%dst(coords%n,coords%d-1))
+  allocate(coords%rdel(coords%n,coords%d))
+  allocate(coords%rdst(coords%n,coords%d-1))
+
+end function allocate_coords
+
+subroutine finalize(this)
+  class(Coords1D), intent(inout) :: this
+  call finalize_Coords1D(this)
+end subroutine finalize
+
+! Deallocate and reset to initial state.
+subroutine finalize_Coords1D(this)
```

Comment:
Was the removal of `guarded_deallocate` a performance enhancement (one less subroutine call) or does it have some other benefit?